### PR TITLE
Parallel benchmarks

### PR DESCRIPTION
**Pull request by class4kayaker**
Fix issue #958.

Testing completed.
For solcx and solkx logs match for single processor, for MPI change in computed error is within epsilon.
For inclusion, script output matches for single processor, and to within a reasonable epsilon for MPI.

### CODE REVIEW

File: benchmark/inclusion/inclusion.cc [hunk 1]
```diff
@@ -480,6 +477,11 @@ namespace aspect
       Vector<float> cellwise_errors_ul2 (this->get_triangulation().n_active_cells());
       Vector<float> cellwise_errors_pl2 (this->get_triangulation().n_active_cells());
 
+      double u_l1;
```
Comment (tjhei) [inline benchmark/inclusion/inclusion.cc]: is there a reason you move the declaration of these variables up here? The code would shorter if you just do
```
double u_l1 = Utilities:: ...
```

File: benchmark/inclusion/run.sh [hunk 2]
```diff
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+NP=1
```
Comment (tjhei) [inline benchmark/inclusion/run.sh]: please add a comment ``# number of processors to run on:``

File: benchmark/solcx/solcx.cc [hunk 3]
```diff
@@ -2773,6 +2770,11 @@ namespace aspect
       Vector<float> cellwise_errors_ul2 (this->get_triangulation().n_active_cells());
       Vector<float> cellwise_errors_pl2 (this->get_triangulation().n_active_cells());
 
+      double u_l1;
```
Comment (tjhei) [inline benchmark/solcx/solcx.cc]: same here

File: benchmark/solkz/solkz.cc [hunk 4]
```diff
@@ -843,6 +840,11 @@ namespace aspect
       Vector<float> cellwise_errors_ul2 (this->get_triangulation().n_active_cells());
       Vector<float> cellwise_errors_pl2 (this->get_triangulation().n_active_cells());
 
+      double u_l1;
```
Comment (tjhei) [inline benchmark/solkz/solkz.cc]: same here

File: benchmark/inclusion/inclusion.cc (hunk 1 above)
Comment (tjhei) [inline benchmark/inclusion/inclusion.cc]: Oh, I realize you just copied this from burstedde.cc. Can you clean it up in that file too please?

**Comment by tjhei** [comment]
Can you change MPI_COMM_WORLD in burstedde.cc to use the communicator too?

**Comment by tjhei** [comment]
Looks good otherwise! :+1:

**Comment by tjhei** [comment]
Can you also update the .cc files in tests/ ? See ``grep n_mpi_processes tests/*cc``. We should also add a test that is run with MPI (maybe copy solcx.prm and run with MPI:2)? Let me know if you need help with that.

**Comment by class4kayaker** [comment]
I have not modified the tangurnis benchmark for parallel due to it outputing data directly to a csv file.

I have updated the tests not based on tangurnis, including the melt tests.

I can start working on the solcx mpi test.

File: tests/melt_compressible_advection.cc [hunk 5]
```diff
@@ -137,8 +134,10 @@ namespace aspect
                                        VectorTools::L2_norm,
                                        &comp_porosity);
 
+    double pore_l2 =  std::sqrt(Utilities::MPI::sum(cellwise_errors_porosity.norm_sqr(),this->get_mpi_communicator()));
```
Comment (tjhei) [inline tests/melt_compressible_advection.cc]: why "pore"? Maybe "poro" is more appropriate?

File: tests/melt_material_4.cc [hunk 6]
```diff
@@ -161,10 +158,15 @@ namespace aspect
                                        quadrature_formula,
                                        VectorTools::L2_norm,
                                        &comp_porosity);
+
+    double u_l2 =  std::sqrt(Utilities::MPI::sum(cellwise_errors_u.norm_sqr(),this->get_mpi_communicator()));
```
Comment (tjhei) [inline tests/melt_material_4.cc]: weird whitespace

File: tests/melt_transport_convergence_simple.cc [hunk 7]
```diff
@@ -196,13 +193,20 @@ namespace aspect
                                        VectorTools::L2_norm,
                                        &comp_u_f);
 
+    double u_l2 =  std::sqrt(Utilities::MPI::sum(cellwise_errors_u.norm_sqr(),this->get_mpi_communicator()));
```
Comment (tjhei) [inline tests/melt_transport_convergence_simple.cc]: weird whitespace

File: tests/melt_transport_convergence_simple.cc [hunk 8]
```diff
@@ -196,13 +193,20 @@ namespace aspect
...
+    double p_l2 =  std::sqrt(Utilities::MPI::sum(cellwise_errors_p.norm_sqr(),this->get_mpi_communicator()));
+    double p_f_l2 =  std::sqrt(Utilities::MPI::sum(cellwise_errors_p_f.norm_sqr(),this->get_mpi_communicator()));
+    double p_c_l2 =  std::sqrt(Utilities::MPI::sum(cellwise_errors_p_c.norm_sqr(),this->get_mpi_communicator()));
+    double pore_l2 =  std::sqrt(Utilities::MPI::sum(cellwise_errors_porosity.norm_sqr(),this->get_mpi_communicator()));
+    double u_f_l2 =  std::sqrt(Utilities::MPI::sum(cellwise_errors_u_f.norm_sqr(),this->get_mpi_communicator()));
```
Comment (bangerth) [inline tests/melt_transport_convergence_simple.cc]: all of these variables can be made `const`

**Comment by tjhei** [comment]
This is ready to be merged, once the tester is through.

**Comment by tjhei** [comment]
please apply the diff from the tester.

**Pull request merged**